tephra-core/src/main/java/co/cask/tephra/persist/HDFSTransactionLog.java

```diff
@@ -157,11 +224,69 @@ public TransactionEdit next(TransactionEdit reuse) throws IOException {
       if (closed) {
         return null;
       }
-      boolean successful = reader.next(key, reuse);
-      if (successful) {
-        return reuse;
+      if (transactionEdits.size() != 0) {
+        return transactionEdits.remove(0);
+      } else {
+        if (newVersion) {
+          // v2 format, buffer edits till we reach a marker
+          boolean successful;
+          DataOutputBuffer rawKey = new DataOutputBuffer();
+
+          if (reader.nextRawKey(rawKey) != -1) {
```

Comment:
Move this in a separate method, combine with the enum usage above, you can organize the code like this:

```java
switch (version) {
  case V2:
    readV2Entries(transactionEdits);
}
...
```